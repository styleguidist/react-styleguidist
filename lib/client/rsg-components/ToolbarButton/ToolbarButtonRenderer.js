import "core-js/modules/es.string.small";
import React from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';
import cx from 'clsx';
export var styles = function styles(_ref) {
  var space = _ref.space,
      color = _ref.color;
  return {
    button: {
      padding: 2,
      // Increase clickable area a bit
      color: color.light,
      background: 'transparent',
      transition: 'color 750ms ease-out',
      cursor: 'pointer',
      '&:hover, &:focus': {
        isolate: false,
        color: color.linkHover,
        transition: 'color 150ms ease-in'
      },
      '&:focus': {
        isolate: false,
        outline: [[1, 'dotted', color.linkHover]]
      },
      '& + &': {
        isolate: false,
        marginLeft: space[1]
      },
      // Style react-icons icon passed as children
      '& svg': {
        width: space[3],
        height: space[3],
        color: 'currentColor',
        cursor: 'inherit'
      }
    },
    isSmall: {
      '& svg': {
        width: 14,
        height: 14
      }
    }
  };
};
export var ToolbarButtonRenderer = function ToolbarButtonRenderer(_ref2) {
  var _cx;

  var classes = _ref2.classes,
      className = _ref2.className,
      onClick = _ref2.onClick,
      href = _ref2.href,
      title = _ref2.title,
      small = _ref2.small,
      testId = _ref2.testId,
      children = _ref2.children;
  var classNames = cx(classes.button, className, (_cx = {}, _cx[classes.isSmall] = small, _cx));

  if (href !== undefined) {
    return /*#__PURE__*/React.createElement("a", {
      href: href,
      title: title,
      className: classNames,
      "aria-label": title,
      "data-testid": testId
    }, children);
  }

  return /*#__PURE__*/React.createElement("button", {
    type: "button",
    onClick: onClick,
    title: title,
    className: classNames,
    "aria-label": title
  }, children);
};
ToolbarButtonRenderer.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
  className: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func,
  title: PropTypes.string,
  small: PropTypes.bool,
  testId: PropTypes.string,
  children: PropTypes.node
};
export default Styled(styles)(ToolbarButtonRenderer);