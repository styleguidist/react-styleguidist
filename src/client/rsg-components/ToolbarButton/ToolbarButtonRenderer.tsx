import React from 'react';
import PropTypes from 'prop-types';
import Styled, { JssInjectedProps } from 'rsg-components/Styled';
import cx from 'clsx';
import * as Rsg from '../../../typings';

export const styles = ({ space, color }: Rsg.Theme) => ({
	button: {
		padding: 2, // Increase clickable area a bit
		color: color.light,
		background: 'transparent',
		transition: 'color 750ms ease-out',
		cursor: 'pointer',
		'&:hover, &:focus': {
			isolate: false,
			color: color.linkHover,
			transition: 'color 150ms ease-in',
		},
		'&:focus': {
			isolate: false,
			outline: [[1, 'dotted', color.linkHover]],
		},
		'& + &': {
			isolate: false,
			marginLeft: space[1],
		},
		// Style react-icons icon passed as children
		'& svg': {
			width: space[3],
			height: space[3],
			color: 'currentColor',
			cursor: 'inherit',
		},
	},
	isSmall: {
		'& svg': {
			width: 14,
			height: 14,
		},
	},
});

interface ToolbarButtonProps extends JssInjectedProps {
	children: React.ReactNode;
	className?: string;
	href?: string;
	onClick?: () => void;
	title?: string;
	small?: boolean;
	testId?: string;
}

export const ToolbarButtonRenderer: React.FunctionComponent<ToolbarButtonProps> = ({
	classes,
	className,
	onClick,
	href,
	title,
	small,
	testId,
	children,
}) => {
	const classNames = cx(classes.button, className, {
		[classes.isSmall]: small,
	});

	if (href !== undefined) {
		return (
			<a href={href} title={title} className={classNames} aria-label={title} data-testid={testId}>
				{children}
			</a>
		);
	}

	return (
		<button type="button" onClick={onClick} title={title} className={classNames} aria-label={title}>
			{children}
		</button>
	);
};

ToolbarButtonRenderer.propTypes = {
	classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	className: PropTypes.string,
	href: PropTypes.string,
	onClick: PropTypes.func,
	title: PropTypes.string,
	small: PropTypes.bool,
	testId: PropTypes.string,
	children: PropTypes.any,
};

export default Styled<ToolbarButtonProps>(styles)(ToolbarButtonRenderer);
