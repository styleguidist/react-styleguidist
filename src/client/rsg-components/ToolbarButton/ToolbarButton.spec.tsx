import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { ToolbarButtonRenderer, styles } from './ToolbarButtonRenderer';

const props = {
	classes: classes(styles),
	title: 'Pizza button',
};

it('should render a button', () => {
	const renderer = createRenderer();
	renderer.render(
		<ToolbarButtonRenderer {...props} onClick={() => {}}>
			pizza
		</ToolbarButtonRenderer>
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should render a link', () => {
	const renderer = createRenderer();
	renderer.render(
		<ToolbarButtonRenderer {...props} href="/foo">
			pizza
		</ToolbarButtonRenderer>
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should pass a class name to a button', () => {
	const renderer = createRenderer();
	renderer.render(
		<ToolbarButtonRenderer {...props} onClick={() => {}} className="foo-class">
			pizza
		</ToolbarButtonRenderer>
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should pass a class name to a link', () => {
	const renderer = createRenderer();
	renderer.render(
		<ToolbarButtonRenderer {...props} href="/foo" className="foo-class">
			pizza
		</ToolbarButtonRenderer>
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should render a button with small styles', () => {
	const renderer = createRenderer();
	renderer.render(
		<ToolbarButtonRenderer {...props} onClick={() => {}} small>
			butterbrot
		</ToolbarButtonRenderer>
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});
