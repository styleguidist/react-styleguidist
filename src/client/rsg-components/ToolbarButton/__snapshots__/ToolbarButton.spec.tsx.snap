// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass a class name to a button 1`] = `
<button
  aria-label="Pizza button"
  className="button foo-class"
  onClick={[Function]}
  title="Pizza button"
  type="button"
>
  pizza
</button>
`;

exports[`should pass a class name to a link 1`] = `
<a
  aria-label="Pizza button"
  className="button foo-class"
  href="/foo"
  title="Pizza button"
>
  pizza
</a>
`;

exports[`should render a button 1`] = `
<button
  aria-label="Pizza button"
  className="button"
  onClick={[Function]}
  title="Pizza button"
  type="button"
>
  pizza
</button>
`;

exports[`should render a button with small styles 1`] = `
<button
  aria-label="Pizza button"
  className="button isSmall"
  onClick={[Function]}
  title="Pizza button"
  type="button"
>
  butterbrot
</button>
`;

exports[`should render a link 1`] = `
<a
  aria-label="Pizza button"
  className="button"
  href="/foo"
  title="Pizza button"
>
  pizza
</a>
`;
